Avoid reducing redirects directly to an Object to avoid memory issue

benoit74 opened the pull request:
Fix #2142 

Reducing the list of redirects directly to an Object caused abnormal memory consumption. While the exact root cause is unknown, reducing first to a Map and then transforming into a Object removes the memory issue. Since Object keys are redirect titles which might be quite big and there might be many in some occasion, we might have memory fragmentation issues or something else.

kelson42 commented:
@benoit74 Do you know when this bug has been introduced?
 
I would appreciate if:
* You check the code base for similar construct (and fix if necessary)
* Put an explanation as comment beside this code so we secure nobody is tempted to reintroduce the "bad" code pattern

Otherwise I'm fine with the PR. I guess it works like you said even this is still a pity we don't understand why we had a memory leak.

benoit74 commented:
> Do you know when this bug has been introduced?

The bug I fixed is in `getArticlesByIds`. Code has been like this since at least 1.5.0.

Code was also similar in `getArticlesByNS` some time ago but it has been changed with https://github.com/openzim/mwoffliner/commit/8461cf03bf542fb9a72b03e812ba1becd37387e1 / https://github.com/openzim/mwoffliner/pull/1728 (released in v1.12.0) for a version which does not exhibit the issue. Commit is huge and PR does not specifically mention the problem. To me it is more accidental that the code has been changed to something which does not exhibit the issue.

I've now looked at the codebase and realized we have a "more common" way to write the code while still not having the memory issue. I wonder if the problem might come from the "unnecessary" `...acc`.  I pushed a new commit with this change instead of previous one.

Looking at the codebase, I found one other `reduce` function with same "weird" code. And many other with the "new" one. Fixed as well in same commit.

I didn't added a comment because it was hard to tell what we should not do, especially since we've not totally understood the issue + it is not the standard "pattern" used in reduce functions.

benoit74 commented:
Nota: there is a non-negligible chance the bug is linked to a Node.JS VM rather than our own code, so it might be more recent than 1.5.0 ... don't know ... and mostly impossible to assess.

kelson42 reviewed (approved):
LGTM, thx for this catch!
